[3.0] Implement name-related improvements identified in previous Vulkan PR

Exanite opened the pull request:
# Summary of the PR

This PR focuses on implementing the naming-related improvements identified in #2457.
Depending on the scope of the tasks, some of these tasks may be pushed to another PR.

This PR also contains the changes of multiple other PRs.
See each individual PR for more information.

Included PRs:
- https://github.com/Exanite/Silk.NET/pull/19
- https://github.com/Exanite/Silk.NET/pull/21
- https://github.com/Exanite/Silk.NET/pull/23
- https://github.com/Exanite/Silk.NET/pull/24
- https://github.com/Exanite/Silk.NET/pull/25
- https://github.com/Exanite/Silk.NET/pull/28
- https://github.com/Exanite/Silk.NET/pull/29

Important changes:
- Handle types are now named as `-HandleEXT` instead of `-EXTHandle`.
   - Eg: `AccelerationStructureHandleKHR`
- Function pointer delegate types are now named as `-DelegateEXT` instead of `EXTDelegate`.
   - Eg: `pfn struct BufferCallbackSOFT` and `pfn delegate BufferCallbackDelegateSOFT`
- Functions with data type suffixes following a number such as `alGetSourcei64vDirectSOFT` are now trimmed and prettified as `GetSourcei64vDirectSOFT` instead of `GetSourcei64VDirectSOFT`. Note the capitalization of `v`. This is because affixes are added after prettification.
- Vendor suffixes are now kept strictly uppercased.
- OpenGL vendor suffix trimming is now generalized and can be configured and used for any Khronos API.
   - Notably, `TrimEnumMemberImpliedVendors` trims enum member vendor suffixes if it matches the enum type vendor suffix.
- Struct properties are now prettified. They were previously ignored.
- NativeName is a new attribute that saves the name output by ClangSharp. This for the most part isn't used by the generator, but can be useful for the end user.
- NameAffix is a new attribute that allows prefixes/suffixes to be declared upfront and reordered. The idea is that mods add the attribute to identifiers and PrettifyNames applies the affixes during trimming and prettification.
- A large portion of the trimming/PrettifyNames code has been rewritten.
   - NameTrimmer has been kept mostly the same.
   - `INameTrimmer.Order` is used to order the trimmers instead of the version.
   - PrettifyNames.Visitor's functionality is mostly the same, but also gathers name affix data.
   - MixKhronosData.Trim is almost all replaced by the name affix system. This is implemented in MixKhronosData.RewriterPhase3 (yes, there are 3 rewriters, but this keeps things simpler without adding much time cost).
   - Most usages of tuples have been replaced with record structs.
   - Most nullable collections (eg: `List<string>?`) have been made non-nullable. Most of these get allocated anyway and it makes the code more maintainable. If there is any performance issue (mainly concerning the Microsoft job), I will take the time to optimize it.
   - Prettification (as in calling the Prettify extension method) is now implemented as another trimmer. Prettification now happens before name conflict resolution and before name affixes are added.
- Handle structs now have constructors since they were previously not possible to construct (field is readonly).
- The StripAttributes mod can be used to remove attributes from the final bindings.
   - Currently the following are removed: `NativeTypeName`, `NameAffix`, `Transformed`

# Related issues, Discord discussions, or proposals

Previous Vulkan PR (initial bindings generation): #2457
Discord thread: https://discord.com/channels/521092042781229087/1376331581198827520/1442651368207941643

This PR also contains the changes of multiple other PRs.
See each individual PR for more information.

Included PRs:
- https://github.com/Exanite/Silk.NET/pull/19
- https://github.com/Exanite/Silk.NET/pull/21
- https://github.com/Exanite/Silk.NET/pull/23
- https://github.com/Exanite/Silk.NET/pull/24
- https://github.com/Exanite/Silk.NET/pull/25
- https://github.com/Exanite/Silk.NET/pull/28
- https://github.com/Exanite/Silk.NET/pull/29

# Further Comments

## Tasks not part of this PR

These are the tasks from my previous PR. These have been sorted and trimmed down to ensure that this PR remains focused. These may be added to this PR.

If you want to see the original, unmodified set of tasks, please see #2457.

Please see this link for an up to date list: https://github.com/orgs/dotnet/projects/64/views/16?pane=issue&itemId=168550672

These will also be part of a future PR.

- [ ] Apply BoolType transformation to VkBool32 in structs. They currently are only handled for functions.
- [ ] Add default field values to structs where it makes sense (eg: `SType`)
- [ ] Ensure SupportedApiProfiles attributes are correct
   - [ ] Properly resolve API profiles for Flags/FlagBits types
   - [ ] Prefer name from `NativeName` for resolving API profiles

Future PRs

- [ ] Update Curin's branch to use my new LocationTransformation code. The new renamer should be replaced with this. See the NameUtils.cs and Renamer.cs files on Curin's branch. Renamer.cs should be removed. NameUtils.cs should have the methods updated to use my LocationTransformation code instead (this should show up as a merge conflict).
   - [x] PR opened and ready for review/merge
   - [x] PR merged
   - [ ] Old renamer code removed
- [ ] Consider opening a ClangSharp PR to fix issue with `SDL_MAX_SINT64` generation on Linux
   - More context: https://discord.com/channels/521092042781229087/1376331581198827520/1443401810822828113
   - `public const nint SDL_MAX_SINT64 = unchecked(0x7FFFFFFFFFFFFFFF);` does not compile
- [ ] Reimplement the struct chaining API
- [ ] Fix issues when running multiple jobs
   - [ ] Figure out why the headers are getting mixed up when running multiple jobs at the same time.
       - Seems to be related to `TransformFunctions`.
   - [ ] Figure out why the SDL handle structs are getting named as `Silk.NET.SDL.ConditionHandle.gen.cs` instead of `ConditionHandle.gen.cs` when running multiple jobs at the same time.
- [ ] Maybe rework how using statements are added. Not sure why, but using statements are randomly added/removed between Windows/Linux. This causes noisy diffs.
- [ ] Consider splitting `ExtractNestedTyping` and `TransformHandles` into a new set of `Extract-` mods.
    - This this mainly for maintainability for `ExtractNestedTyping`, but for `TransformHandles` it is useful for `AddApiProfiles` to be executed after all types are extracted, but not yet transformed.
    - Also consider making `AddApiProfiles` work off of `NativeName` attributes.
- [ ] Optimize/refactor name overrides in PrettifyNames
- [ ] Consider reworking data type trimming regexes.
   - This was discussed more here: https://github.com/dotnet/Silk.NET/pull/2503#issuecomment-3622907776
   - The summary is that the current set of 2 regexes doesn't elegantly handle the case where the ending of a word should not be trimmed, but the word itself can have data type suffixes.
   - For example: `alSourceRewindv` and `alSourceRewind`. The `d` in `Rewind` can be mistaken as a data type suffix, so we add it to the `EndingsNotToTrim` regex. However, the `dv` in `Rewindv` also has the same problem.
   - We likely need something more akin to `WordsNotToTrimInto` than `EndingsNotToTrim`.
- [ ] Optimize `NativeName` attributes.
   -  This should involve renaming `StripAttributes` to `CleanupAttributes` and adding an option in its config to optimize `NativeName` attributes.
   - Perksey seems to really want this so considering prioritizing this: https://github.com/dotnet/Silk.NET/pull/2503#discussion_r2662921224
- [ ] Consider replacing `LenientUnderscore` and updating codebase to use `NameSplitter`
- [ ] Consider implementing a `IsProbablyABitmask` function based on [BuildTool's IsProbablyABitmask](https://github.com/dotnet/Silk.NET/blob/c87fc664cc5d15abd754997b0a9ad138470554c4/src/Core/Silk.NET.BuildTools/Common/Utilities.cs#L627) for identifying flags enums
   - Original comment: https://github.com/dotnet/Silk.NET/pull/2503#discussion_r2662888928
- [ ] Improve handling of nested struct names
	- Eg: `PerformanceCounterDescriptionArmName` should be `PerformanceCounterDescriptionARMName`
	- Conversation here: https://discord.com/channels/521092042781229087/587346162802229298/1460017726813900975
- [ ] Address `INameTrimmer.Order/Version` issue (we don't want magic numbers) and consider removing `INameTrimmer` entirely.

Done during later PRs:

- [x] Prevent `OES` being singularizing into `O` in `ArrayParameterTransformer.TransformArrayParameterRewriter`.
   - Eg: `glFeedbackBufferxOES` 
   - Discussed solution is to expose the name affix API so that `TransformArrayParameterRewriter` can singularize the base name instead of the full name.
- [x] Change native name of `GLEnum` to `GLenum`. Same for `ALCenum` and `ALenum`.

## Completed Todos

Newest groups at the top. Order within a group is chronological.

Note that I also wrote a summary at the top. The summary is more comprehensive, but this section can be useful getting additional context. Also see my commit descriptions if more context is needed.

Prepare to merge.

- [x] Finish self code review
- [x] Generate bindings on Windows for consistency
- [x] Temporarily remove generated files from PR during code review
   - This needs to be done since there are too many generated files and they hide some of the manually changed files. 
   - [x] Revert the commit after code review: https://github.com/Exanite/Silk.NET/pull/19
- [x] Review the generated bindings as a second code review
- [x] Consider updating OpenAL submodule and regenerating
- [x] Merge name affix config PR into this PR
- [x] Fix broken test cases (I know there is at least one)

Acronym threshold change and prettify improvements

- [x] Acronym related name improvements discussed here: https://discord.com/channels/521092042781229087/587346162802229298/1446979314330501172
   - [x] Re-read discussion and organize into tasks.
   - [x] Change default acronym threshold to 2 for all bindings
   - [x] Change default acronym threshold to 2 to be the default if not specified in the PrettifyNames config.
   - [x] Handle `Mat4X4` case. The "X" should be lowercased.
   - [x] Update existing names to match framework design guidelines
      - [x] `EGL` as `Egl`
      - [x] Name overrides in generator config

INameTrimmer.Order

- [x] Add priority system to `INameTrimmer`.
   - I implemented the new name affixer system using trimmers, so this introduced 3 new trimmers in order to separate each stage properly. There is now a new `INameTrimmer.Order` property and internal `TrimmerOrder` enum.

Fix misc issues I noticed

- [x] `VkVendorId` members are not trimmed properly
- [x] Figure out why the members of the enum `AttribMask` in OpenGL are not getting rewritten.
   - Initial investigation shows that the `fieldSymbol.ConstantValue` is null, but why it is null is unclear. Maybe a Roslyn bug? Example member: `DepthBufferBit = unchecked((uint)0x00000100)`. Note that similar expressions in Vulkan are fine, eg: `unchecked((ulong)0x00000002UL)`. Maybe it is because the literal type and cast type are different.
   - Wait, are these the enums that we generate?
   - Yup, they are. MixKhronosData was outputting subtly invalid type names. I changed it to use `ParseTypeName(baseType)` instead of `IdentifierName(baseType)`.

Fix issues introduced by name affix system

- [x] Identify and fix issues introduced by name affixation. Also just identify and fix naming issues in general.
   - [x] Data type trimming in OpenAL seems to be broken. Probably OpenGL as well. 
   - [x] Check if I need to reimplement vendor suffix removal for OpenGL. I probably do.
      - [x] Discuss: I decided to change the vendor suffix removal for enum members so that it removes the member suffix if it matches the type suffix. This prevents cases where a member of a different suffix is added, causing all members to have their suffixes restored.
      - [x] Discuss: The above change made it so that the member suffix removal code is usable for Vulkan so I enabled it for Vulkan as well. See PresentModeKHR to see how it looks (the KHR suffixed members have their suffixes removed, but EXT remains the same).
   - [x] Reapply some of my `MixKhronosData` code removals to the current `develop/3.0` branch and see what changes would be made. This is to ensure I don't accidentally remove functionality when deleting code.
- [x] Ensure other APIs are configured consistently with Vulkan. Eg: Ensure affix priorities are set, etc.

Name affix system

- [x] Properly trim prefixed/suffixed names, such as `PFNVkDebugUtilsMessengerCallbackEXT` and `BufferTHandle`
   - [x] `PFNVkDebugUtilsMessengerCallbackEXT` prefix
   - [x] `PFNVkDebugUtilsMessengerCallbackDelegateEXT` suffix
   - [x] `AccelerationStructureHandleKHR` suffix
   - [x] Extract vendor extensions from names and store into `NameSuffix` attributes.
   - [x] Revert "hacky" THandle fix.
- [x] Consider renaming handle types as `HandleEXT` instead of `EXTHandle` for readability.
   - This is fairly doable. The above todos track this. 

Name metadata

- [x] Add NativeName attribute
- [x] Ensure NativeName attribute is added by all mods that generate new types/members
   - Technically not strictly required, but nice from a API consistency standpoint. 
   - [x] MixKhronosData
      - [x] Enum members
      - [x] Enum declarations
         - [x] Vulkan (uses FlagBits version of enum names since it is more helpful when googling)
         - [x] Discuss validity in other Khronos APIs (some APIs don't have actual enums)
            - Eg: `GLEnum` and the other OpenGL enums
   - [x] ExtractNestedTyping
      - [x] Function pointers
      - [x] Others?
   - [x] TransformHandles 

Name prettification

- [x] ~~Change naming convention to not capitalize acronyms (i.e., strictly pascal cased)~~
   - ~~See: https://discord.com/channels/521092042781229087/1376331581198827520/1395097379329282089~~ 
   - [x] ~~Update generated method class names in the rsp files to be strictly pascal case~~
- [x] Fix `StdVideoEncodeH264SliceHeaderFlags` property names. They currently aren't getting prettified.

Fix bugs from original PR

- [x] Enums containing negative values should not have an unsigned backing type.

Handle struct improvements

- [x] Handle structs don't have an easy way of being constructed (field is readonly and no constructor).

Exanite committed:
Remove acronym identification in order to make all names be strictly pascal cased

Exanite committed:
Allow multiple NameAffix attributes to be used on one identifier

Exanite committed:
Cleanup

Exanite committed:
Don't process secondaries when adding to outputing to newNames dictionary

The secondaries are never used so this doesn't affect anything and simplifies the code.
I'm guessing the TODO comment implies that these were

Exanite committed:
Fix incorrect rename

Exanite committed:
Regenerate all bindings on Linux after changing acronyms to be pascal cased

Lots of breaking changes and also a good amount of non-ideal changes. These non-ideal changes will be fixed later in this PR.

Exanite committed:
Update acronyms in names to be pascal cased and fix compile errors

Exanite committed:
Remove NameUtils.Prettify allowAllCaps parameter

Exanite committed:
Remove transformer from NameUtils.Prettify API and prevent unnecessary allocation by using an preallocated array

Exanite committed:
Properly prettify the prettify only names if no trimmer baseline is set and refactor code

Exanite committed:
Cleanup code

Exanite committed:
Refactor name affix API

Exanite committed:
Implement name affix handling

Note that running the generator with this commit unmodified will lead to to a crash in NameUtils.Prettify due to an empty string input.
This is because of VkVendorId. You'll have to replace the exception in NameUtils.Prettify with a return statement (just return `ret`) to have the generator work.

Exanite committed:
Add prefix override for VkMemoryDecompressionMethodFlagsEXT

Exanite committed:
Improve NameUtils.Prettify error messages

Exanite committed:
Add VendorSuffixIdentifierExclusions to MixKhronosData config

Exanite committed:
Regenerate SDL and OpenAL bindings. OpenGL needs more work. SDL has no change.

Exanite committed:
Remove now unnecessary _T trimming code

Exanite committed:
Remove unused code

Exanite committed:
Remove ExtensionVendorTrimmingMode

It's not used and using it causes issues

Exanite committed:
Add todos

Exanite committed:
Add NameAffixerLateTrimmer back and use prettify prevention hack to get OpenGL bindings to generate properly

All 4 non-Microsoft APIs were regenerated in this commit.

Exanite committed:
Merge branch 'feature/vulkan-bindings-3.0-name-affixer' into feature/vulkan-bindings-3.0-improvements

Exanite committed:
Fix a few compile errors

Exanite committed:
Handle method vendor suffixes and remove vendor suffix capitalization hacks used in MixKhronosData

Exanite committed:
Add NativeName attribute for function pointer structs and delegates

Exanite committed:
Ensure delegate vendor extensions are identified properly

Exanite committed:
Cleanup

Exanite committed:
Update generator configs for other jobs and also strip NameAffix attributes

Exanite committed:
Add TransformEnums mod to all jobs

Exanite committed:
Ignore cast expressions when deciding whether to rewrite enum members and handle case where EqualsValue or ConstantValue is null

The first change was targeted towards the OpenGL bindings (see AttribMask), but it looks like ConstantValue is null for those as well.

Exanite committed:
Regenerate SDL (inconsistent generation?)

Exanite committed:
Fix non strict pascal casing in generator config

Does not affect final output, but does affect the intermediate name

Exanite committed:
Prevent struct method names from getting trimmed

Exanite committed:
Rename VendorSuffixIdentifierExclusions to VendorSuffixIdentificationExclusions

Exanite committed:
Fix compile error

Exanite committed:
Update comments

Exanite committed:
Add AttributeUtils.GetNativeNameOrDefault()

Exanite committed:
Attempt to reimplement exclusive vendor suffix trimming code. Results currently greatly differ for non-OpenGL APIs.

Exanite committed:
Remove the now unnecessary rewind code

This also fixes the generation issue that VendorId has

Exanite committed:
Temporarily remove data type trimming code

Exanite committed:
Fix ALC_INVALID_CONTEXT being prettified as InvalidContEXT

Exanite committed:
Organize configuration for MixKhronosData and pre-emptively add new options that will be used by later commits

Exanite committed:
Add configuration options for TrimEnumTypeNonExclusiveVendors and TrimEnumMemberSameVendors

Exanite committed:
Replace TrimEnumMemberExclusiveVendors with TrimEnumMemberImpliedVendors and enable for Vulkan

Exanite committed:
Decide not to special case OpenGL's SamplePatternSGIS

Exanite committed:
Optimize/cleanup trimming code

Exanite committed:
Attempt to implement data type trimming, but not working yet due to failing name conflict resolution

Exanite committed:
Merge branch 'feature/vulkan-bindings-3.0-improvements-enum-vendor-trimming' into feature/vulkan-bindings-3.0-improvements

Exanite committed:
Edit doc comment for NameAffix

Exanite committed:
Add discriminator priority to NameAffix attribute

Exanite committed:
Add ToString implementations for CandidateNames structs for easier debugging

Exanite committed:
Add comment explaining PreventPrettificationHack

Exanite committed:
Avoid adding redundant secondaries

Exanite committed:
Merge branch 'feature/vulkan-bindings-3.0-improvements' into feature/vulkan-bindings-3.0-data-type-trimming

Exanite committed:
Regen OpenAL on Linux

Exanite committed:
Change secondaries to be non-nullable since I allocate all of them anyway. Will profile and optimize later.

Exanite committed:
Regen OpenGL on Linux

Exanite committed:
Disable stripping for NameAffix attributes in OpenAL for easier debugging

Exanite committed:
Rework the NameAffix properties and what they do again

Exanite committed:
Update config options and adjust docs a bit more

Exanite committed:
Edit docs

Exanite committed:
Store the NameAffixAttribute declaration order

Exanite committed:
Update order and priority of data type suffixes

Exanite committed:
Attempt to get data trimming working and apparently fix a few bugs in the process

Exanite committed:
Fix AL_VOCAL_MORPHER_PHONEME_T's -T suffix getting trimmed

Exanite committed:
Edit doc comment on NameAffixerEarlyTrimmer.Version

Exanite committed:
Regenerate all bindings on Linux

Exanite committed:
Strip all NameAffix attributes for OpenAL since I'm no longer debugging

Exanite committed:
Sort and ensure consistent attribute targets for NativeName and NameAffix attributes

Exanite committed:
Fix formatting

Exanite commented on sources/Core/Core/Annotations/NameAffixAttribute.cs:
```diff
@@ -0,0 +1,66 @@
+    /// <para/>
+    /// Affixes with the same priority are applied together as a group.
+    /// </summary>
+    public int Priority { get; }
```
Consider renaming Order and Priority since this is confusing.

Exanite committed:
Don't mark parameters with the NativeName attribute

Exanite committed:
Add INameTrimmer.Order property and sort trimmers by that instead of version

Exanite committed:
Fix OpenAL's alSourceRewindv getting trimmed as alSourceRewin

Exanite committed:
Fix incorrect range indexing

Exanite committed:
Change prettification to occur before affixes are re-added

Exanite committed:
Rename NameAffix.Priority to DiscriminatorPriority for clarity

Exanite committed:
Cleanup RewriterPhase3 code in MixKhronosData

Exanite reviewed (commented):
Initial self code review. PR is not yet ready for review or merge.

Exanite committed:
Fix incorrect hasSetter condition causing properties to be incorrectly trimmed

Exanite committed:
Fix TexStorageAttribs and ContainerType getting over trimmed

Also, apparently I didn't regenerate OpenGL following some of my previous changes. The data type suffixes in particular have changed.

Exanite committed:
Fix enums created by MixKhronosData not having their enum values rewritten properly

Exanite committed:
Regenerate bindings on Windows

Exanite committed:
Fix compile errors in tests

Exanite committed:
Exclude SDL_SetWindowsMessageHook

Exanite committed:
Exclude SDL_FILE, SDL_LINE, SDL_NULL_WHILE_LOOP_CONDITION

Exanite committed:
Fix compile errors and update windowing code to use strict pascal case for GL/EGL

Exanite committed:
Revert after review: Revert all generated bindings changes

Exanite committed:
Revert incorrect OpenGL capitalization changes in SdlContext

Exanite committed:
Revert incorrect OpenGL capitalization changes in SdlSurfaceComponents.OpenGL.cs

Exanite reviewed (commented):
Self code review complete. Will mark as ready for review shortly.

Note that I only have the non-generated changes committed.

The rest is in a separate commit that I will apply after the initial review.
This is because the generated changes hide the non-generated changes (it goes over the 1k file limit).

DO NOT merge the PR until the generated changes are committed, and preferably, reviewed.

Exanite commented:
Ready for review!

If you want to see the generated code see: https://github.com/dotnet/Silk.NET/pull/2503/commits/9c9fbfc037b662a260ff5c8eccbdeda4c064cc78.

Note that the red/removed code is the new code since this code is meant to temporarily revert the generated changes.

Exanite committed:
Revert "Revert after review: Revert all generated bindings changes"

This reverts commit 9c9fbfc037b662a260ff5c8eccbdeda4c064cc78.

Exanite committed:
Begin reverting acronym casing change

Exanite committed:
Update openal-soft submodule

Exanite committed:
Revert change to EGLPlatformInfo

Exanite committed:
Regenerate bindings on Windows

Exanite committed:
Revert after review: Revert all generated bindings changes

Exanite committed:
Revert changes to ContextSourceExtensions

Exanite committed:
Revert changes to SdlEventProcessor

Exanite committed:
Revert changes to SdlSurfaceComponents

Exanite committed:
Revert "Revert after review: Revert all generated bindings changes"

This reverts commit 9bbf8be3e65f2dc56fb6a6b8722bfe172dc33c60.

Exanite commented on sources/SilkTouch/SilkTouch/Mods/PrettifyNames.cs:
```diff
@@ -336,26 +323,100 @@ z.MethodKind is MethodKind.Constructor or MethodKind.Destructor
+    /// Then if no override is found, then the name's affixes are removed,
+    /// the name is prettified, and the name's affixes are added back.
+    /// </summary>
+    private string ApplyPrettifyOnlyPipeline(
```
This is from the Microsoft branch, but heavily modified.

Exanite commented on sources/SilkTouch/SilkTouch/Mods/PrettifyNames.cs:
```diff
@@ -336,26 +323,100 @@ z.MethodKind is MethodKind.Constructor or MethodKind.Destructor
+        ICulturedStringTransformer nameTransformer)
+    {
+        // Check for overrides
+        foreach (var (nativeName, overriddenName) in nameOverrides)
```
Ideally this code isn't duplicated with the code in Trim(), but to do this well requires a decent amount of refactoring.

Perksey replied:
Worth doing at some point though.

Exanite reviewed (commented):
Self code review complete. Going to mark as ready for review.

Note that I only have the non-generated changes committed.
The generated changes are here: https://github.com/Exanite/Silk.NET/pull/19

This is because the generated changes hide the non-generated changes (it goes over the 1k file limit).

Feel free to comment on the linked PR as well.
It will be merged into this PR after the non-generated changes are reviewed.

DO NOT merge this PR until the generated changes are merged, and preferably, reviewed.

Perksey commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -2140,6 +1828,190 @@ private bool TryGetManagedEnumType(SyntaxList<AttributeListSyntax> attributes, [
+
+        public override SyntaxNode VisitEnumDeclaration(EnumDeclarationSyntax node)
+        {
+            // Special case for enums since this code needs information about the enum type and its members at the same time.
```
Why does it need to know its type?

Perksey commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -2140,6 +1828,190 @@ private bool TryGetManagedEnumType(SyntaxList<AttributeListSyntax> attributes, [
+                return false;
+            });
+
+            var isSafeToTrimMembers = !containsUnsuffixed;
```
Why does it not containing any unsuffixed vendors make it safe? How is this different to exclusiveVendor?

Perksey replied:
The new comments explain this much better, thanks!

Perksey commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -2140,6 +1828,190 @@ private bool TryGetManagedEnumType(SyntaxList<AttributeListSyntax> attributes, [
+                {
+                    if (isSafeToTrimType)
+                    {
+                        // Remove the exclusive vendor from the enum name since it is wrong and it is safe to do so
```
Why is it safe to do so? Why is it wrong? Can you give an example of where this code is used? This could result in breaking changes!

Perksey replied:
The comments help. And yeah, you're right, this does introduce breaking changes in that if you have `BufferTargetARB` in one version of the spec, and update to one that for some reason has `BufferTargetNV`, then you've got an API break because this code will revive the suffix for `BufferTargetARB` after it had previously been trimmed to `BufferTarget`.

But given that OpenGL is the only case where I expect this will be used, I'm not too worried about it. That was the case I was worried about though, generally I strive to make our API maximally resistant to changes like these.

Perksey commented on sources/SilkTouch/SilkTouch/Naming/INameTrimmer.cs:
```diff
@@ -15,6 +15,12 @@ public interface INameTrimmer
+    /// Used to define the order that the trimmers will run in.
+    /// Higher values indicate that the trimmer should run later.
+    /// </summary>
+    int Order { get; }
```
I don't think this should be the way we determine the order of execution - this is not something individual augments should be able to influence without careful consideration of what it's trying to achieve. If the order of execution matters and has different effects, this should be user configurable. As I've said before, I want to get rid of things like `Version` where you arbitrarily change the number to achieve the result you want - if there is a need to control the order of execution to achieve the results the user wants, then that's what the user-facing design should reflect.

Exanite replied:
The only external trimmer (external as in not part of PrettifyNames) here is MixKhronosData.
This means we currently don't really have a usage pattern to analyze.

What other trimmers will there be that isn't covered by the existing functionality?
Even `NameTrimmer` can probably use the name affix system instead (as a `IdentifyGlobalPrefixes` mod that has a syntax rewriter), which has a well defined and configurable execution order.

MixKhronosData does provide one case, which is to prevent/rewind trimming in the case of `GLEnum`.
Generally speaking, this would be more of a name replacement/transformation. This is not covered by the affix system, but is partially covered by the name override system.

Exanite commented on sources/SilkTouch/SilkTouch/Mods/MarkNativeNames.cs:
```diff
@@ -0,0 +1,113 @@
+                return attributeLists;
+            }
+
+            return attributeLists.WithNativeName(identifier.Text);
```
This should be something we do as a post-process step and not at a pre-process step.

MarkNativeNames is designed as a pre-process step. It goes early in the mod order.
MarkNativeNames goes right after ClangScraper because, at that point, we know that all names match the native names (outside some exceptions like in bitfield structs where ClangSharp introduces an extra field).

Attribute removal is currently handled by `StripAttributes`, which is a post-process step. It goes late (last currently) in the mod order.
This mod can be generalized as an `CleanupAttributes` mod where optimizing attributes is part of the configurable steps that the mod applies.

The reason why optimizing attributes should be done as a post-process is because optimization requires knowledge of the output. If we introduce the "don't add if redundant" logic during pre-processing, then no `NativeName` attributes would ever get added.

An alternative is for mods to only add the `NativeName` attribute if they change the name, but that also requires all mods to understand the rules for adding/removing the `NativeName` attribute, which greatly introduces complexity.

Perksey replied:
Hmmmmm ok, I agree - I must admit I haven't checked the bindings yet so I don't know whether this is being done already.

Exanite replied:
The optimization of NativeName attributes isn't, but it's not too hard to implement.
To me, this is a lower priority task since it's not critical to the bindings working.

Exanite commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -2140,6 +1828,190 @@ private bool TryGetManagedEnumType(SyntaxList<AttributeListSyntax> attributes, [
+
+            if (config.TrimEnumTypeNonExclusiveVendors && typeVendor != null)
+            {
+                var shouldTrimType = typeVendor != exclusiveVendor;
```
This code mostly matches the functionality of the code that was originally in `MixKhronosData.Trim`.
Well... to the best of my ability since that code wasn't really commented either.

The exceptions are that I removed/prevented some edge cases and restructured the code.

I provide a better explanation of the functionality of the code in the `Configuration` record struct at the top of the class, but I agree that that this code is likely hard to understand without context and local examples. I'll go through the code and add these.

Perksey commented:
Oh... lol. Ok yeah maybe this is the appropriate fix.

Perksey commented on generator.json:
```diff
@@ -300,13 +361,22 @@
+        "ExcludeVendorSuffixIdentification": [
+          "VK_VENDOR_ID_VIV",
+          "VK_VENDOR_ID_VSI",
+          "VK_VENDOR_ID_MESA"
```
Good spot.

Exanite commented:
As in, combine the two regexes? I think that would make it confusing to edit.

Edit: I think I see what you mean. You mean to make it so that the regex considers if trimming the suffix would trim into part of another word. Eg: Match for `Rewind` before matching for `dv` or `v`.

Perksey commented:
I'm concerned about only having Rewind in EndingsToTrim if that is indeed the valid fix. I still feel like there should be a way to use EndingsNotToTrim to get what we want but I'm not sure how. But I don't want the two regexes to not enact the correct behaviour because they have different sets of applicable words.

Exanite committed:
Add Rewindv to EndingsNotToTrim

Exanite committed:
Add Rewindv to EndingsNotToTrim

(cherry picked from commit 167ef888d574ad29584a6ad38d338a2f6d444c03)

Exanite committed:
Add more comments on enum vendor suffix trimming

(cherry picked from commit 933e1bcb4662a26ae0ccb8f0f5f158589ca65b93)

Exanite committed:
Add more comments on enum vendor suffix trimming

Exanite committed:
Add example for isSafeToTrimType and move condition to a named variable for clarity

Exanite committed:
Add example for isSafeToTrimType and move condition to a named variable for clarity

(cherry picked from commit 87aed8e767b270a1c991632adbe49efa77d79ec1)

Exanite committed:
Add comment shouldTrimType

Exanite committed:
Add comment on intended use of Order and DiscriminatorPriority properties

Exanite committed:
Merge branch 'feature/vulkan-bindings-3.0-improvements' into feature/vulkan-bindings-3.0-improvements-unrevert

Exanite commented:
For the record, the PR is ready to review again, but I'm currently doing some additional reviewing of the generated bindings to ensure that I haven't changed anything that I didn't mean to change.

Exanite committed:
Implement new name affix category approach

Exanite committed:
Add validation for name affix category configs

Exanite committed:
Adjust API of AddNamePrefix/Suffix extensions and add more docs

Exanite committed:
Update existing name affix identification code to use new API

Exanite committed:
Update enum vendor suffix identification comments to be consistent with new affix identification approach

Exanite committed:
Shorten PrettifyNames.NameAffixes config option to just Affixes

Exanite committed:
Fix incorrect default value for DiscriminatorPriority

Exanite committed:
Shorten name of DiscriminatorPriority configuration option

Exanite committed:
Update generator.json to use new name affix config

Exanite committed:
Add name affix attributes for -Internal (intercepted) and -Raw functions

Exanite committed:
Fix incorrect parameter order (oops)

Exanite committed:
Configure RawFunction and InterceptedFunction categories

Exanite committed:
Redesign config and begin updating implementation to support new config

Exanite committed:
New name affix config compiles, but logic has not been updated

Exanite committed:
Update name affix configs in generator.json

Exanite committed:
Update name affix removal/application code to work with new config

Exanite committed:
Fix incorrect sort order

Exanite committed:
Add temporary hack to fix -OES being trimmed as -O issue

Exanite committed:
Merge remote-tracking branch 'dotnet/Silk.NET/develop/3.0' into feature/name-affixes-categories-update

Exanite committed:
Regenerate on Windows and remove old cache files

Perksey reviewed (commented):
That review felt like a breeze, thanks so much for your hard work!

Exanite commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -1887,53 +1602,42 @@ private class RewriterPhase1(JobData job, ILogger logger) : CSharpSyntaxRewriter
+                                                                EnumMemberDeclaration(x.Identifier.ToString())
+                                                                    .WithAttributeLists(
+                                                                        new SyntaxList<AttributeListSyntax>()
+                                                                        .WithNativeName(x.Identifier.Text))
```
Ah, that's good to know. I'll look into running it for all of the files I've modified in this PR.

Perksey commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -1807,12 +1520,14 @@ public bool TryGetSymbolMetadata(
             + "Fixed|Pass|Address|Configs|Thread|Subpass|Deferred|Extended|Affix|Annex|Box|Aux|Ex|Index|Vertex|Path|"
             + "Arch|Arith|Afresh|Both|High|Math|Mesh|Sinh|Bench|Brush|Bunch|Crash|Flush|Depth|Latch|Morph|Pinch|"
-            + "Pitch|Stretch|Smooth|Matrix|Radix|Sound|Rewind|Supported)$"
+            + "Pitch|Stretch|Smooth|Matrix|Radix|Sound|Supported|Rewind|Rewindv)$"
```
Awesome, thanks. I suspect there's something a little bit fundamental about the approach e.g. maybe we should trim just one data type at a time and compare with `EndingsNotToTrim` after every trim. Little things to play with later.

Exanite committed:
Reorder KhronosVendor suffixes to be last

Exanite committed:
Merge branch 'develop/3.0' into feature/csharpier-files-modified-by-my-pr

Exanite committed:
Format all files changed in my name improvements set of PRs

Command used (csharpier v0.29.2):
dotnet csharpier sources/Core/Core/Annotations/NameAffixAttribute.cs &&
dotnet csharpier sources/Core/Core/Annotations/NativeNameAttribute.cs &&
dotnet csharpier sources/Core/Core/Annotations/SupportedApiProfileAttribute.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Clang/WindowsStdIncludeResolver.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/JobDependencies.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/AddIncludes.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/AddVTables.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Bakery/DefaultBakeStrategy.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Common/AttributeUtils.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Common/ModCSharpSyntaxRewriter.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Common/ModLoader.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Common/ModUtils.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/ExtractNestedTyping.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/InterceptNativeFunctions.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/MarkNativeNames.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/PrettifyNames.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/StripAttributes.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/TransformEnums.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/TransformHandles.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Transformation/ArrayParameterTransformer.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Mods/Transformation/FunctionTransformer.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/CandidateNames.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/CandidateNamesWithOriginal.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/INameTrimmer.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/NameTrimmer217.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/NameTrimmer218.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/NameTrimmerContext.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/NameUtils.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Naming/TrimmerOrder.cs &&
dotnet csharpier sources/SilkTouch/SilkTouch/Sources/IInputResolver.cs &&
dotnet csharpier tests/SilkTouch/SilkTouch/Khronos/MixKhronosDataTests.cs &&
dotnet csharpier tests/SilkTouch/SilkTouch/Naming/NameTests.cs

Exanite committed:
Format again by rerunning exact same command as previous commit

Not sure why the output is unstable, but it's fine after this second execution.

Exanite committed:
Rewrite NameAffix arg parsing code to look better under CSharpier formatting

Exanite committed:
Merge pull request #25 from Exanite/feature/csharpier-files-modified-by-my-pr

[3.0] Format files I changed using CSharpier

Exanite committed:
Merge pull request #24 from Exanite/feature/khronos-vendor-last

[3.0] Reorder KhronosVendor suffixes to be last

Exanite committed:
Merge pull request #23 from Exanite/feature/name-affixes-categories-update

[3.0] Update feature/name-affixes-categories with changes from develop/3.0

Exanite committed:
Merge pull request #21 from Exanite/feature/name-affix-categories

[3.0] Implement new name affix category approach

Exanite committed:
Merge pull request #19 from Exanite/feature/vulkan-bindings-3.0-improvements-unrevert

[3.0] Regenerate bindings on Windows for Name Improvements PR

Perksey reviewed (approved):
Approved once marked ready for review... lol.

Exanite committed:
Remove affixes before global prefix identification; Change NameTrimmer to trim the primaries instead of the original name

Exanite committed:
Add DummyJobDependency and DummyOptions for use during testing

Exanite committed:
Add 3 unit tests to replace OverzealousNameTrimming test case and commit the test that already passes

Exanite committed:
Add test case: MixKhronosData_IdentifiesImpliedVendorSuffixes

Exanite committed:
Add test case: MixKhronosData_IdentifiesNonExclusiveVendorSuffixes

Exanite committed:
Add new TestUtils class with TestUtils.CreateTestProject

Exanite committed:
Add test case: PrettifyNames_TrimsSharedPrefix_AfterRemovalOf_VendorSuffixes_AndShortenedNamesConflict

Exanite committed:
Remove unused null assertion

Exanite committed:
Simplify tests and apparently MixKhronosData should read the config from the jobData not the options snapshot

Exanite committed:
Move CandidateNamesWithOriginal to NameTrimmer and rename to AugmentedCandidateNames; Replace the dictionary used with a list since it was a glorified list of tuples

I might change it back to be a dictionary later, but using a list is safer for now since it ensures that all accesses need to directly access a property of AugmentedCandidateNames.

This change was originally to fix an issue where trimming names collide after affixes are removed.
This is an issue introduced by my change to the INameTrimmer order (see f33d5922c09e31c4d33c56477deb9d7efe126900) since I made it so that affixes are now removed before NameTrimmer runs.

Exanite committed:
Verify result of PrettifyNames_TrimsSharedPrefix_AfterRemovalOf_VendorSuffixes_AndShortenedNamesConflict

Exanite committed:
Replace OverzealousNameTrimmingFixupIsNotOverzealousForOpenAL test case with PrettifyNames_TrimsSharedPrefix2

Exanite committed:
Move PrettifyNames tests to a new file

Exanite committed:
Simplify and cleanup PrettifyNames tests

Exanite committed:
Fix syntax error in test code

Exanite committed:
Shorten test names

Exanite committed:
Fix incorrect document name

Exanite committed:
Add PrettifyNames tests: TrimsPrefix_WhenMatchingHint and DoesNotTrimTypeName_WhenNotMatchingHint_AndOnlyOneType

Exanite committed:
Fix incorrect comment

Exanite committed:
Add default value for LongAcronymThreshold and change type from int? to int

Exanite committed:
Add PrettifyNames test: TrimsSharedPrefix_ForTypes

Exanite committed:
Add tests: HintShouldNotAffectSharedPrefixTrimming and HintShouldNotAffectSharedPrefixTrimming_WhenAffixesDeclared

Exanite committed:
Dictionary is no longer nullable

Exanite committed:
Use Append instead of Concat+Repeat

Exanite committed:
Edit comment for clarity

Exanite committed:
Replace use of ternary since it makes the code really annoying to debug

Exanite committed:
Fix DoesNotTrimTypeName_WhenNotMatchingHint_AndOnlyOneType

Exanite committed:
Remove unnecessary affix since it doesn't affect whether the test fails or not

Exanite committed:
Fix TrimsSharedPrefix_WhenAffixesDeclared and HintShouldNotAffectSharedPrefixTrimming_WhenAffixesDeclared

Exanite committed:
Add NameUtilsTests and rename NameTests to NameTrimmerTests

Exanite committed:
Add failing test: Prettify_IsNotAffectedBy_TrailingUnderscore

Exanite committed:
Fix Prettify_IsNotAffectedBy_TrailingUnderscore by trimming input

The reason why is that Humanize() in NameUtils.Prettify() stops breaking up the name if the name ends in a trailing underscore.
This is probably intended on Humanizer's part, but is not the behavior we want.

Exanite committed:
Rename and accept result of Regression_UnexpectedCasingChangesInFormatEnums; the new capitalized name is correct

Exanite committed:
Add tests: Prettify_Capital_AfterNumber_DoesNotAffect_PreviousWord and InconsistentCasing_LettersFollowingNumbers_WhenAffixesDeclared

Exanite committed:
Prefilter matches for null or whitespace to avoid possibility of any weird logic bugs

Exanite committed:
Rename matches to words for clarity

Exanite committed:
Add expected result of InconsistentCasing_LettersFollowingNumbers_WhenAffixesDeclared

Exanite committed:
Add new PrettifyTests file and move Prettify-related tests to it

Exanite committed:
Change Prettify to be strictly for C# identifiers

Exanite committed:
Move Prettify to be next to NameTransformer since they strictly work together

Exanite committed:
Add set of tests: PrettifyTests.CommonCases

Exanite committed:
Redesign Prettify API and make NamePrettifier its own standalone class

Exanite committed:
Add more tests

Exanite committed:
Add more tests

(cherry picked from commit 4827fa975bd50ea2c1d7979dbcfc9d4289497bd4)

Exanite committed:
Add more test cases for Prettify

Exanite committed:
Merge branch 'feature/mixkhronosdata-tests' into feature/prettify-rewrite

Exanite committed:
Add test Lowercase_AfterNumber_IsPartOf_PreviousWord

Exanite committed:
Rename PrettifyTests to NamePrettifierTests

Exanite committed:
Attempt 1 at writing word breaker

Mostly works, but decided I want to try another approach.

Exanite committed:
Implement new word splitter

Exanite committed:
Cleanup

Exanite committed:
Add word fragment merging

Exanite committed:
Add two more prettify tests

Exanite committed:
Add more test cases

Exanite committed:
Fix incorrect test cases

Exanite committed:
Implement new prettifier (almost fully working)

Exanite committed:
Remove debug and placeholder code

Exanite committed:
Update expected result of PrettifyNamesTests.Regression_UnexpectedCasingChangesInFormatEnums

Exanite committed:
Use try-finally to update tracking variable; uhh kinda cursed but this might read better

Exanite committed:
Change IsAcronym to also check the word length

Exanite committed:
Accept new changes to PrettifyNamesTests.InconsistentCasing_LettersFollowingNumbers_WhenAffixesDeclared

Exanite committed:
Accept changes to PrettifyNamesTests.TrimsSharedPrefix (this is correct, but non-ideal; the NameAffix code handles this case separately)

Exanite committed:
Clarify comments

Exanite committed:
Handle case where the identifier is all caps

Exanite committed:
Fix incorrect test cases caused by bad assumptions

Exanite committed:
Organize NamePrettifier tests and add comments

Exanite committed:
Fix a few regressions

Also currently trying to ensure old behavior before changing things, so the number x number change has been reverted for now.

Exanite committed:
Add NameSplitterTests

Exanite committed:
Temporarily disable acronym conflict resolution

Exanite committed:
Remove old code

Exanite committed:
Add more NameSplitterTests

Exanite committed:
Apply the stdint remap to all current bindings

Exanite committed:
Exclude ALC_CPLUSPLUS

Exanite committed:
Change expected output for lowercase fragments so that they are capitalized like normal words

Exanite committed:
Add test: Regression_IncorrectSecondary_ChosenAsFallback

Exanite committed:
Fix PrettifyNamesTests.Regression_IncorrectSecondary_ChosenAsFallback and accept output

Exanite committed:
Also exclude AL_CPLUSPLUS

Exanite committed:
Change MixKhronosData to identify the GL- prefix of GLEnum instead of using INameTrimmer.Trim to restore it after it gets trimmed

Exanite committed:
Add test case for splitting SpvImageFormatR32ui

Exanite committed:
Add UpperCaseFirstCharacter utility function

Exanite committed:
Short all uppercase names are now finally handled properly?

Exanite committed:
Rename Regressions test case to AcronymsWithNumbers

Exanite committed:
Reorder tests

Exanite committed:
Add comment

Exanite committed:
Also remap GLintptrARB and GLsizeiptrARB

This fixes inconsistencies in the generated bindings between Linux/Windows and makes the types match the promoted core versions since the promoted core versions are already being remapped.

Exanite committed:
Add docs to new NamePrettifier and cleanup

Exanite committed:
Ignore lowercase "x" when checking if name is all non-lowercase

Exanite committed:
Regenerate on Windows (part 1 due to case insensitivity)

Exanite committed:
Regenerate on Windows (part 2 due to case insensitivity)

Exanite committed:
Add more NamePrettifier tests

Exanite committed:
Adjust acronym threshold for N3D AcronymsWithNumbers test case

Exanite committed:
Fix invalid tests

Exanite committed:
Add another test case

Exanite committed:
Lowercase "X" if it is between two numbers

Exanite committed:
Implement acronym conflict checking and add more tests

Exanite committed:
Regenerate on Linux since Windows case insensitivity sucks

Exanite committed:
Regenerate on Windows

Exanite committed:
Move name splitting and char categorization code out of NamePrettifier

Exanite committed:
Add todo for TrimsSharedPrefix_ForTypes test case

Exanite committed:
Fix PrettifyNamesTests.TrimsSharedPrefix_ForTypes

Exanite committed:
Clean up my prefix trimming fix

Exanite committed:
Change default LongAcronymThreshold to 2

Exanite committed:
Generate on Linux to avoid Windows case insensitivity issues

Exanite committed:
Regenerate on Windows

Exanite committed:
Replace EGL with Egl where it makes sense to do so

Exanite committed:
Add remark comment explaining why the default is 2

Exanite committed:
Update expected test outputs

These changes are expected because the default acronym threshold is now 2 and the suffixes are not declared as NameAffixes.

Exanite committed:
Fix casing of NumAabBs by overriding it

Exanite committed:
Don't scope the NameOverride for numAABBs

Exanite committed:
Revert incorrect comment change (these refer to native types outside of our control)

Exanite committed:
Merge pull request #29 from Exanite/feature/acronym-threshold

[3.0] Change acronym threshold to 2 for all bindings

Exanite committed:
Merge pull request #28 from Exanite/feature/prettify-rewrite

[3.0] Heavily expand testing suite; Rewrite NameUtils.Prettify; Fix a lot of subtle naming bugs

Perksey commented on sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:
```diff
@@ -1800,7 +1505,7 @@ public bool TryGetSymbolMetadata(
     /// </summary>
     // NOTE: LET THIS BE A LESSON! Do NOT add x (fixed) here, these will frequently conflict with integer overloads.
-    [GeneratedRegex("(?<!xe)([fdh]v?|u?[isb](64)?v?|v|i_v|fi|hi)$")]
+    [GeneratedRegex("(?<!xe)((?<!Rewin)[dhf]v?|u?[isb](64)?v?|v|i_v|fi|hi)$")]
```
Would like this investigated in a future PR. (Pretty sure it's already on the todo list)
